### PR TITLE
feat: Implement fast-deep-equal faster

### PR DESCRIPTION
**Pull request by streamich**
Tweaks the algo to make it faster.

Below are results of 3x benchmark run:

```
fast-deep-equal-2 x 127,075 ops/sec ±2.32% (81 runs sampled)
fast-deep-equal x 106,086 ops/sec ±1.07% (87 runs sampled)
nano-equal x 88,697 ops/sec ±1.65% (76 runs sampled)
shallow-equal-fuzzy x 73,826 ops/sec ±1.11% (86 runs sampled)
underscore.isEqual x 53,620 ops/sec ±1.55% (82 runs sampled)
lodash.isEqual x 23,102 ops/sec ±1.24% (86 runs sampled)
deep-equal x 30,485 ops/sec ±1.97% (81 runs sampled)
deep-eql x 15,491 ops/sec ±1.22% (85 runs sampled)
assert.deepStrictEqual x 361 ops/sec ±2.42% (80 runs sampled)
The fastest is fast-deep-equal-2
```

```
fast-deep-equal-2 x 135,825 ops/sec ±1.88% (85 runs sampled)
fast-deep-equal x 126,627 ops/sec ±2.73% (85 runs sampled)
nano-equal x 109,734 ops/sec ±1.09% (90 runs sampled)
shallow-equal-fuzzy x 95,265 ops/sec ±0.70% (88 runs sampled)
underscore.isEqual x 67,421 ops/sec ±0.99% (90 runs sampled)
lodash.isEqual x 27,459 ops/sec ±4.50% (79 runs sampled)
deep-equal x 35,484 ops/sec ±1.22% (91 runs sampled)
deep-eql x 19,298 ops/sec ±1.10% (83 runs sampled)
assert.deepStrictEqual x 449 ops/sec ±2.35% (85 runs sampled)
The fastest is fast-deep-equal-2
```

```
fast-deep-equal-2 x 141,601 ops/sec ±1.03% (89 runs sampled)
fast-deep-equal x 137,028 ops/sec ±0.88% (90 runs sampled)
nano-equal x 111,036 ops/sec ±0.97% (86 runs sampled)
shallow-equal-fuzzy x 93,458 ops/sec ±1.54% (89 runs sampled)
underscore.isEqual x 68,208 ops/sec ±0.74% (93 runs sampled)
lodash.isEqual x 30,475 ops/sec ±0.86% (90 runs sampled)
deep-equal x 35,877 ops/sec ±0.85% (91 runs sampled)
deep-eql x 19,368 ops/sec ±0.73% (91 runs sampled)
assert.deepStrictEqual x 447 ops/sec ±1.52% (83 runs sampled)
The fastest is fast-deep-equal-2
```

### CODE REVIEW

**Comment by epoberezkin** [comment]
@streamich cool, thank you. will you make this PR to update the actual code?

**Comment by streamich** [comment]
Merged it into the actual code.

File: index.js [hunk 1]
```diff
@@ -1,42 +1,51 @@
...
-    if (dateA != dateB) return false;
+  var regexpA = a instanceof RegExp
+    , regexpB = b instanceof RegExp;
+  if (regexpA != regexpB) return false;
 
-    var regexpA = a instanceof RegExp
-      , regexpB = b instanceof RegExp;
-    if (regexpA && regexpB) return a.toString() == b.toString();
-    if (regexpA != regexpB) return false;
+  if (dateA && dateB) return a.getTime() == b.getTime();
```
Comment (epoberezkin) [inline index.js]: I was wondering, why this line is not immediately after line 27?

File: index.js [hunk 2]
```diff
@@ -1,42 +1,51 @@
...
-    for (i = 0; i < keys.length; i++)
-      if (!Object.prototype.hasOwnProperty.call(b, keys[i])) return false;
+  if (a instanceof Object && b instanceof Object) {
+    var keys = keyList(a);
+    length = keys.length;
 
-    for (i = 0; i < keys.length; i++)
-      if(!equal(a[keys[i]], b[keys[i]])) return false;
+    for (i = 0; i < length; i++)
+      if (!(keys[i] in b)) return false;
```
Comment (epoberezkin) [inline index.js]: I understand it may be faster but "in" is a very tricky operator - it always looks up the prototype chain so there can be surprises. I would stick with Object.prototype.hasOwnProperty.call - maybe it should be assigned to a var as well.

File: index.js [hunk 3]
```diff
@@ -1,42 +1,51 @@
...
+    length = a.length;
+    if (length != b.length) return false;
+    for (i = 0; i < length; i++)
       if (!equal(a[i], b[i])) return false;
     return true;
   }
 
   if (arrA != arrB) return false;
 
-  if (a && b && typeof a === 'object' && typeof b === 'object') {
```
Comment (epoberezkin) [inline index.js]: I would also add tests for "null" - removing a && b from the condition could let null pass maybe? Although `null instanceof Object` is false...

**Comment by streamich** [comment]
I have removed `in` keyword, and ran the benchmark 3 times:

```
fast-deep-equal-new x 139,648 ops/sec ±1.32% (87 runs sampled)
fast-deep-equal-original x 133,418 ops/sec ±1.36% (90 runs sampled)
```

```
fast-deep-equal-new x 140,931 ops/sec ±0.72% (88 runs sampled)
fast-deep-equal-original x 132,433 ops/sec ±1.96% (88 runs sampled)
```

```
fast-deep-equal-new x 141,231 ops/sec ±0.63% (88 runs sampled)
fast-deep-equal-original x 135,583 ops/sec ±1.48% (87 runs sampled)
```

File: index.js (hunk 2 above)
Comment (streamich) [reply index.js]: Removed `in` keyword.

File: index.js (hunk 1 above)
Comment (streamich) [reply index.js]: Because it is probably faster to do `instanceof` than `.getTime()`, also `.toString()` on regular expression is definitely slow.

Comment (epoberezkin) [reply index.js]: but it would only do `.getTime()` if it should do it (both and an b are `instanceof Date`), and in this case there is no need to `instanceof RegExp` at all.

**Comment by epoberezkin** [comment]
so still faster :)

**Comment by streamich** [comment]
Moved

```js
if (dateA && dateB) return a.getTime() == b.getTime();
```

up

File: index.js [hunk 4]
```diff
@@ -1,42 +1,51 @@
...
 
-    for (i = 0; i < keys.length; i++)
-      if(!equal(a[keys[i]], b[keys[i]])) return false;
+    for (i = 0; i < length; i++) {
+      key = keys[i];
+      if (!equal(a[key], b[key])) return false;
+    }
 
-    return true;
+    return length == keyList(b).length;
```
Comment (epoberezkin) [inline index.js]: sorry, I can be missing something, but shouldn't this condition be on line 39 as `if (length != keyList(b).length) return false`? It looks like the simplest and the fastest inequality check - if the number of keys is different it should immediately return false, rather than comparing all key values first.

Comment (streamich) [reply index.js]: I think `keyList(b).length` is very expensive, and when moving it to the top results drop.

The reason is that `keyList` creates a new array, which we postpone to the very end, only if necessary.

For, example if `b` is

```
{
  a: 1,
  b: 1,
  b: 1,
}
```

`keyList` has to create an array for us

```js
['a', 'b', 'c']
```

just so we can look at its `.length` property.

I am actually thinking now, maybe I should iterate through the object myself and count the number of keys in each object instead of doing `keyList(a).length` and `keyList(b).length`.

Comment (epoberezkin) [reply index.js]: I think the results will depend on the size of the object. Checking and comparing the number of keys can be expensive, but it's predictably expensive and should scale linearly with the number of keys. Comparing the values in keys can be unpredictably slow as they can have very deep data hierarchies. So I think we should compare the number of keys as the first thing and then either iterate and compare key names (as is now) or it may be faster to call equal(keysArray1, keysArray2) - not sure, probably extra function call will make slower.

**Comment by streamich** [comment]
@epoberezkin I've mode the key length check before the recursive call, the new version still is faster with current benchmark:

```
fast-deep-equal x 134,145 ops/sec ±1.15% (89 runs sampled)
fast-deep-equal-original x 130,394 ops/sec ±0.90% (91 runs sampled)

fast-deep-equal x 143,424 ops/sec ±0.77% (91 runs sampled)
fast-deep-equal-original x 137,625 ops/sec ±1.40% (86 runs sampled)

fast-deep-equal x 146,275 ops/sec ±0.66% (91 runs sampled)
fast-deep-equal-original x 139,929 ops/sec ±0.79% (91 runs sampled)
```
